src/traits.jl

```diff
@@ -129,13 +129,16 @@ eltype(::Type{Colorant{T,N}}) where {T,N}   = T
 
 eltype(c::Colorant) = eltype(typeof(c))
 
+@pure function _parameter_upper_bound(t::UnionAll, idx)
```

Comment:
It can be risky to claim a function is supposedly `@pure`. This was deprecated in Base due to this issue, and because the type system can now handle such queries correctly through dispatch(*), while `similar` already implements what's often meant, but better.

(*) The mistake often made in using this function is the assumption that subtype parameters are of the same number and order as their supertype. This happens to be true inside this package for some of its types, but is easily seen to be false for other examples like `BitArray <: AbstractArray{Bool}`. Thus, the correct solution is usually to ask dispatch for the matching parameters of the _supertype_, like so:
```
eltype(::Type{<:AbstractArray{T}}) where {T} = T
```

Reply:
> The mistake often made in using this function is the assumption that subtype parameters are of the same number and order as their supertype. 

I understand. However, that assumption seems necessary. I think it is the correct behavior to get the wrong result with the wrong usage.